webhook/stats_reporter.go

```diff
@@ -71,37 +71,77 @@ type StatsReporter interface {
 	ReportConversionRequest(request *apixv1.ConversionRequest, response *apixv1.ConversionResponse, d time.Duration) error
 }
 
+type options struct {
+	tagsToExclude map[string]struct{}
+}
+
+type Option func(_ *options)
+
+func WithoutTag(tag string) Option {
+	return func(opts *options) {
+		if opts.tagsToExclude == nil {
+			opts.tagsToExclude = make(map[string]struct{})
+		}
+		opts.tagsToExclude[tag] = struct{}{}
+	}
+}
+
 // reporter implements StatsReporter interface
 type reporter struct {
-	ctx context.Context
+	ctx  context.Context
+	opts options
 }
 
 // NewStatsReporter creates a reporter for webhook metrics
-func NewStatsReporter() (StatsReporter, error) {
+func NewStatsReporter(opts ...Option) (StatsReporter, error) {
 	ctx, err := tag.New(
 		context.Background(),
 	)
 	if err != nil {
 		return nil, err
 	}
 
-	return &reporter{ctx: ctx}, nil
+	options := options{}
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &reporter{ctx: ctx, opts: options}, nil
 }
 
 // Captures req count metric, recording the count and the duration
 func (r *reporter) ReportAdmissionRequest(req *admissionv1.AdmissionRequest, resp *admissionv1.AdmissionResponse, d time.Duration) error {
-	ctx, err := tag.New(
-		r.ctx,
-		tag.Insert(requestOperationKey, string(req.Operation)),
-		tag.Insert(kindGroupKey, req.Kind.Group),
-		tag.Insert(kindVersionKey, req.Kind.Version),
-		tag.Insert(kindKindKey, req.Kind.Kind),
-		tag.Insert(resourceGroupKey, req.Resource.Group),
-		tag.Insert(resourceVersionKey, req.Resource.Version),
-		tag.Insert(resourceResourceKey, req.Resource.Resource),
-		tag.Insert(resourceNamespaceKey, req.Namespace),
-		tag.Insert(admissionAllowedKey, strconv.FormatBool(resp.Allowed)),
-	)
+	mutators := []tag.Mutator{}
+
+	if _, ok := r.opts.tagsToExclude[requestOperationKey.Name()]; !ok {
```

Comment:
Could we make this more compact? For example:

```
"k8s.io/apimachinery/pkg/util/sets"
...
	keySet := sets.Set[string]{}
	tagforKey := map[string]struct {
		key   tag.Key
		value string
	}{
		requestOperationKey.Name():  {key: requestOperationKey, value: string(req.Operation)},
		kindGroupKey.Name():         {key: kindGroupKey, value: req.Kind.Group},
		kindVersionKey.Name():       {key: kindVersionKey, value: req.Kind.Version},
		kindKindKey.Name():          {key: kindKindKey, value: req.Kind.Kind},
		resourceGroupKey.Name():     {key: resourceGroupKey, value: req.Resource.Group},
		resourceVersionKey.Name():   {key: resourceVersionKey, value: req.Resource.Version},
		resourceResourceKey.Name():  {key: resourceResourceKey, value: req.Resource.Resource},
		resourceNamespaceKey.Name(): {key: resourceNamespaceKey, value: req.Namespace},
		admissionAllowedKey.Name():  {key: admissionAllowedKey, value: strconv.FormatBool(resp.Allowed)},
	}
	for key := range tagforKey {
		keySet.Insert(key)
	}
	keysToKeep := keySet.Difference(r.opts.tagsKeysToExclude)
	for key := range keysToKeep {
		mutators = append(mutators, tag.Insert(tagforKey[key].key, tagforKey[key].value))
	}

```
that means you need to adjust:

```
type options struct {
	tagsKeysToExclude sets.Set[string]
}
```

Reply:
Good suggestions. I updated it again as almost exactly like you suggested.